infobind/cache.go

```diff
@@ -0,0 +1,225 @@
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	rs := tmp[0]
+	defer terror.Call(rs.Close)
+
+	fs := rs.Fields()
+	chkBatch := rs.NewRecordBatch()
+	for {
+		err = rs.Next(context.TODO(), chkBatch)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		fmt.Println("chkBatch num row():", chkBatch.NumRows())
+		if chkBatch.NumRows() == 0 {
+			return nil
+		}
+		it := chunk.NewIterator4Chunk(chkBatch.Chunk)
+		for row := it.Begin(); row != it.End(); row = it.Next() {
+			record := decodeBindTableRow(row, fs)
+			err = bc.appendNode(h.ctx, record, h.parser)
+			if err != nil {
+				continue
+			}
+
+			if record.UpdateTime.Compare(h.lastUpdateTime) == 1 {
+				h.lastUpdateTime = record.UpdateTime
+			}
+		}
+
+		chkBatch.Reset()
+	}
+}
+
+// Update update the HandleUpdater's bindCache if tidb first startup,the fullLoad is true,otherwise fullLoad is false.
+func (h *HandleUpdater) Update(fullLoad bool) error {
+	var (
+		err error
+		sql string
+	)
+	bc := h.globalHandle.Get()
+
+	newBc := &bindCache{
+		Cache: make(map[string][]*bindData, len(bc.Cache)),
+	}
+
+	for hash, bindDataArr := range bc.Cache {
+		newBc.Cache[hash] = append(newBc.Cache[hash], bindDataArr...)
+	}
+
+	if fullLoad {
+		sql = fmt.Sprintf("select original_sql, bind_sql, default_db, status, create_time, update_time, charset, collation from mysql.bind_info")
+	} else {
+		sql = fmt.Sprintf("select original_sql, bind_sql, default_db, status, create_time, update_time, charset, collation from mysql.bind_info where update_time > \"%s\"", h.lastUpdateTime.String())
+	}
+	err = h.loadDiff(sql, newBc)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	h.globalHandle.bind.Store(newBc)
+
+	fmt.Println("newBc len:", len(newBc.Cache))
+	return nil
+}
+
+func parseSQL(parser *parser.Parser, sql string, charset string, collation string) ([]ast.StmtNode, []error, error) {
+	return parser.Parse(sql, charset, collation)
+}
+
+func decodeBindTableRow(row chunk.Row, fs []*ast.ResultField) bindRecord {
+	var value bindRecord
+
+	value.OriginalSQL = row.GetString(0)
+	value.BindSQL = row.GetString(1)
+	value.Db = row.GetString(2)
+	value.Status = row.GetInt64(3)
+	value.CreateTime = row.GetTime(4)
+	value.UpdateTime = row.GetTime(5)
+	value.Charset = row.GetString(6)
+	value.Collation = row.GetString(7)
+
+	return value
+}
+
+func (b *bindCache) appendNode(sctx sessionctx.Context, newBindRecord bindRecord, sparser *parser.Parser) error {
+	hash := parser.DigestHash(newBindRecord.OriginalSQL)
+
+	if bindArr, ok := b.Cache[hash]; ok {
+		for idx, v := range bindArr {
+			if v.OriginalSQL == newBindRecord.OriginalSQL && v.Db == newBindRecord.Db {
+				b.Cache[hash] = append(b.Cache[hash][:idx], b.Cache[hash][idx+1:]...)
+				break
+			}
+		}
+	}
+
+	// If the bindRecord has been deleted, the status will be 0 or will be 1.
+	if newBindRecord.Status == 0 {
+		if _, ok := b.Cache[hash]; ok {
+			if len(b.Cache[hash]) == 0 {
```

Comment:
I think we can move this check to line 196 and only needs to check if the `len(b.cache[hash])`.

Reply:
OK